File: backend/skills/environment.go
package skills

import (
	"os"
	"strings"
)

var knownEnvironments = map[string]bool{
	"kanban": true,
	"docker": true,
	"s6":     true,
}

var envDetectCache = make(map[string]bool)

func detectEnvironment(env string) bool {
	if val, ok := envDetectCache[env]; ok {
		return val
	}

	result := true
	switch env {
	case "kanban":
		if os.Getenv("ENOUGH_KANBAN_TASK") != "" || os.Getenv("ENOUGH_KANBAN_BOARD") != "" ||
			os.Getenv("HERMES_KANBAN_TASK") != "" || os.Getenv("HERMES_KANBAN_BOARD") != "" {
			result = true
		} else {
			result = false
		}
	case "docker":
		result = isContainer()
	case "s6":
		if fi, err := os.Stat("/run/s6"); err == nil && fi.IsDir() {
			result = true
		} else if fi, err := os.Stat("/package/admin/s6-overlay"); err == nil && fi.IsDir() {
			result = true
		} else {
			result = false
		}
	}

	envDetectCache[env] = result
	return result
}

func isContainer() bool {
	if _, err := os.Stat("/.dockerenv"); err == nil {
		return true
	}
	if _, err := os.Stat("/run/.containerenv"); err == nil {
		return true
	}
	data, err := os.ReadFile("/proc/1/cgroup")
	if err == nil {
		content := string(data)
		if strings.Contains(content, "docker") || strings.Contains(content, "podman") || strings.Contains(content, "/lxc/") {
			return true
		}
	}
	return false
}

func SkillMatchesEnvironment(fm map[string]interface{}) bool {
	envVal, ok := fm["environments"]
	if !ok || envVal == nil {
		return true
	}

	var list []string
	switch v := envVal.(type) {
	case string:
		if v != "" {
			list = []string{v}
		}
	case []string:
		list = v
	case []interface{}:
		for _, item := range v {
			if s, ok := item.(string); ok {
				list = append(list, s)
			}
		}
	}

	if len(list) == 0 {
		return true
	}

	for _, env := range list {
		normalized := strings.ToLower(strings.TrimSpace(env))
		if normalized == "" {
			continue
		}
		if !knownEnvironments[normalized] {
			return true
		}
		if detectEnvironment(normalized) {
			return true
		}
	}

	return false
}

File: backend/skills/environment.ts
// PORT: backend/skills/environment.go

import fs from "node:fs";

const knownEnvironments: Record<string, boolean> = {
  kanban: true,
  docker: true,
  s6: true,
};

const envDetectCache = new Map<string, boolean>();

function detectEnvironment(env: string): boolean {
  if (envDetectCache.has(env)) {
    return envDetectCache.get(env)!;
  }

  let result = true;
  switch (env) {
    case "kanban":
      if (
        (process.env.HOLLOW_KANBAN_TASK !== undefined && process.env.HOLLOW_KANBAN_TASK !== "") ||
        (process.env.HOLLOW_KANBAN_BOARD !== undefined && process.env.HOLLOW_KANBAN_BOARD !== "") ||
        (process.env.HERMES_KANBAN_TASK !== undefined && process.env.HERMES_KANBAN_TASK !== "") ||
        (process.env.HERMES_KANBAN_BOARD !== undefined && process.env.HERMES_KANBAN_BOARD !== "")
      ) {
        result = true;
      } else {
        result = false;
      }
      break;
    case "docker":
      result = isContainer();
      break;
    case "s6":
      try {
        const fi = fs.statSync("/run/s6");
        if (fi.isDirectory()) {
          result = true;
        } else {
          result = false;
        }
      } catch {
        try {
          const fi = fs.statSync("/package/admin/s6-overlay");
          if (fi.isDirectory()) {
            result = true;
          } else {
            result = false;
          }
        } catch {
          result = false;
        }
      }
      break;
  }

  envDetectCache.set(env, result);
  return result;
}

function isContainer(): boolean {
  try {
    fs.statSync("/.dockerenv");
    return true;
  } catch {}
  try {
    fs.statSync("/run/.containerenv");
    return true;
  } catch {}
  try {
    const data = fs.readFileSync("/proc/1/cgroup", "utf8");
    if (data.includes("docker") || data.includes("podman") || data.includes("/lxc/")) {
      return true;
    }
  } catch {}
  return false;
}

export function SkillMatchesEnvironment(fm: Record<string, any>): boolean {
  const envVal = fm["environments"];
  if (envVal === undefined || envVal === null) {
    return true;
  }

  let list: string[] = [];
  if (typeof envVal === "string") {
    if (envVal !== "") {
      list = [envVal];
    }
  } else if (Array.isArray(envVal)) {
    for (const item of envVal) {
      if (typeof item === "string") {
        list.push(item);
      }
    }
  }

  if (list.length === 0) {
    return true;
  }

  for (const env of list) {
    const normalized = env.trim().toLowerCase();
    if (normalized === "") {
      continue;
    }
    if (!knownEnvironments[normalized]) {
      return true;
    }
    if (detectEnvironment(normalized)) {
      return true;
    }
  }

  return false;
}

/*
PORT STATUS
source path: backend/skills/environment.go
source lines: 104
draft lines: 105
confidence: high
status: phase_b_compile
*/
